File: backend/secureattend-backend/src/main/java/com/secureattend/service/AuthService.java
package com.secureattend.service;

import com.secureattend.domain.*;
import com.secureattend.dto.*;
import com.secureattend.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import java.util.Optional;
import java.util.UUID;

@Service
public class AuthService {
    
    @Autowired
    private PersonRepository personRepository;
    
    @Autowired
    private CourseClassRepository courseClassRepository;
    
    @Autowired
    private PasswordEncoder passwordEncoder;


    public AuthResponse professorSignup(SignupRequest request) {
        Optional<Person> existing = personRepository.findByEmail(request.getEmail());
        if (existing.isPresent()) {
            throw new RuntimeException("Email already registered");
        }

        Professor professor = new Professor();
        professor.setId(UUID.randomUUID().toString());
        professor.setName(request.getName());
        professor.setEmail(request.getEmail());
        professor.setPassword(passwordEncoder.encode(request.getPassword()));
        professor.setRole(Person.Role.PROFESSOR);
        professor.setDepartment(request.getDepartment());
        professor.setActive(true);

        personRepository.save(professor);
        
        String token = generateToken(professor);
        return new AuthResponse(token, professor.getId(), professor.getName(), 
                                "PROFESSOR", professor.getEmail());
    }

    public AuthResponse professorLogin(LoginRequest request) {
        Person person = personRepository.findByEmail(request.getUsername())
            .orElseThrow(() -> new RuntimeException("Invalid email or password"));

        if (!passwordEncoder.matches(request.getPassword(), person.getPassword())) {
            throw new RuntimeException("Invalid email or password");
        }

        if (person.getRole() != Person.Role.PROFESSOR) {
            throw new RuntimeException("Not authorized as professor");
        }

        String token = generateToken(person);
        return new AuthResponse(token, person.getId(), person.getName(), 
                                "PROFESSOR", person.getEmail());
    }

    public Student enrollStudent(EnrollStudentRequest request) {
        CourseClass courseClass = courseClassRepository.findById(request.getClassId())
            .orElseThrow(() -> new RuntimeException("Class not found"));

        Optional<Person> existing = personRepository.findByStudentNumber(request.getRollNumber());
        Student student;

        if (existing.isPresent()) {
            Person person = existing.get();
            if (!(person instanceof Student)) {
                throw new RuntimeException("Roll number is already used by a non-student user");
            }
            
            student = (Student) person;
            
            if (request.getFaceImageBase64() != null && !request.getFaceImageBase64().isEmpty()) {
                student.setFaceImageBase64(request.getFaceImageBase64());
                personRepository.save(student);
            }
            
            if (courseClass.getStudentIds().contains(student.getId())) {
                return student;
            }
            
        } else {
            student = new Student();
            student.setId(UUID.randomUUID().toString());
            student.setName(request.getName());
            student.setStudentNumber(request.getRollNumber());
            student.setPassword(passwordEncoder.encode(request.getPassword()));
            student.setRole(Person.Role.STUDENT);
            student.setEmail(request.getEmail());
            student.setMajor(request.getMajor());
            student.setYear(request.getYear());
            if (request.getFaceImageBase64() != null && !request.getFaceImageBase64().isEmpty()) {
                student.setFaceImageBase64(request.getFaceImageBase64());
            }
            student.setActive(true);

            personRepository.save(student);
        }

        if (!courseClass.getStudentIds().contains(student.getId())) {
            courseClass.getStudentIds().add(student.getId());
            courseClassRepository.save(courseClass);
        }

        return student;
    }

    public AuthResponse studentLogin(LoginRequest request) {
        Person person = personRepository.findByStudentNumber(request.getUsername())
            .orElseThrow(() -> new RuntimeException("Invalid roll number or password"));

        if (!passwordEncoder.matches(request.getPassword(), person.getPassword())) {
            throw new RuntimeException("Invalid roll number or password");
        }

        if (person.getRole() != Person.Role.STUDENT) {
            throw new RuntimeException("Not authorized as student");
        }

        String token = generateToken(person);
        return new AuthResponse(token, person.getId(), person.getName(), 
                                "STUDENT", person.getEmail());
    }

    public TA createTA(CreateTARequest request) {
        Optional<Person> existing = personRepository.findByTaId(request.getTaId());
        if (existing.isPresent()) {
            throw new RuntimeException("TA ID already exists");
        }

        TA ta = new TA();
        ta.setId(UUID.randomUUID().toString());
        ta.setName(request.getName());
        ta.setTaId(request.getTaId());
        ta.setPassword(passwordEncoder.encode(request.getPassword()));
        ta.setRole(Person.Role.TA);
        ta.setEmail(request.getEmail());
        ta.setDepartment(request.getDepartment());
        ta.setSupervisorProfessorId(request.getSupervisorProfessorId());
        ta.setActive(true);
        
        if (request.getProfilePictureBase64() != null && !request.getProfilePictureBase64().isEmpty()) {
            ta.setProfilePictureBase64(request.getProfilePictureBase64());
        }

        personRepository.save(ta);
        return ta;
    }

    public AuthResponse taLogin(LoginRequest request) {
        Person person = personRepository.findByTaId(request.getUsername())
            .orElseThrow(() -> new RuntimeException("Invalid TA ID or password"));

        if (!passwordEncoder.matches(request.getPassword(), person.getPassword())) {
            throw new RuntimeException("Invalid TA ID or password");
        }

        if (person.getRole() != Person.Role.TA) {
            throw new RuntimeException("Not authorized as TA");
        }

        String token = generateToken(person);
        return new AuthResponse(token, person.getId(), person.getName(), 
                                "TA", person.getEmail());
    }

    private String generateToken(Person person) {
        return "TOKEN_" + person.getId() + "_" + System.currentTimeMillis();
    }

    public Person validateToken(String token) {
        if (token == null || !token.startsWith("TOKEN_")) {
            throw new RuntimeException("Invalid token");
        }

        String[] parts = token.split("_");
        if (parts.length < 3) {
            throw new RuntimeException("Invalid token format");
        }

        String userId = parts[1];
        return personRepository.findById(userId)
            .orElseThrow(() -> new RuntimeException("User not found"));
    }
}

File: backend/secureattend-node/src/services/authService.js
import bcrypt from 'bcryptjs';
import { randomUUID } from 'crypto';
import { User } from '../models/User.js';
import { CourseClass } from '../models/CourseClass.js';

function generateToken(user) {
  return `TOKEN_${user._id}_${Date.now()}`;
}

function toAuthResponse(user, token) {
  return {
    token,
    userId: user._id,
    name: user.name,
    role: user.role,
    email: user.email || '',
  };
}

export async function professorSignup(request) {
  const existing = await User.findOne({ email: request.email });
  if (existing) throw new Error('Email already registered');

  const professor = new User({
    _id: randomUUID(),
    name: request.name,
    email: request.email,
    password: await bcrypt.hash(request.password, 10),
    role: 'PROFESSOR',
    department: request.department,
    active: true,
  });
  await professor.save();

  const token = generateToken(professor);
  return toAuthResponse(professor, token);
}

export async function professorLogin(request) {
  const person = await User.findOne({ email: request.username });
  if (!person || !(await bcrypt.compare(request.password, person.password))) {
    throw new Error('Invalid email or password');
  }
  if (person.role !== 'PROFESSOR') throw new Error('Not authorized as professor');

  const token = generateToken(person);
  return toAuthResponse(person, token);
}

export async function enrollStudent(request) {
  const courseClass = await CourseClass.findById(request.classId);
  if (!courseClass) throw new Error('Class not found');

  let student = await User.findOne({ studentNumber: request.rollNumber });

  if (student) {
    if (student.role !== 'STUDENT') {
      throw new Error('Roll number is already used by a non-student user');
    }
    if (request.faceImageBase64) {
      student.faceImageBase64 = request.faceImageBase64;
      await student.save();
    }
    if (courseClass.studentIds.includes(student._id)) {
      return student.toJSON();
    }
  } else {
    student = new User({
      _id: randomUUID(),
      name: request.name,
      studentNumber: request.rollNumber,
      password: await bcrypt.hash(request.password, 10),
      role: 'STUDENT',
      email: request.email,
      major: request.major,
      year: request.year,
      faceImageBase64: request.faceImageBase64 || undefined,
      active: true,
    });
    await student.save();
  }

  if (!courseClass.studentIds.includes(student._id)) {
    courseClass.studentIds.push(student._id);
    await courseClass.save();
  }

  return student.toJSON();
}

export async function studentLogin(request) {
  const person = await User.findOne({ studentNumber: request.username });
  if (!person || !(await bcrypt.compare(request.password, person.password))) {
    throw new Error('Invalid roll number or password');
  }
  if (person.role !== 'STUDENT') throw new Error('Not authorized as student');

  const token = generateToken(person);
  return toAuthResponse(person, token);
}

export async function createTA(request) {
  const existing = await User.findOne({ taId: request.taId });
  if (existing) throw new Error('TA ID already exists');

  const ta = new User({
    _id: randomUUID(),
    name: request.name,
    taId: request.taId,
    password: await bcrypt.hash(request.password, 10),
    role: 'TA',
    email: request.email,
    department: request.department,
    supervisorProfessorId: request.supervisorProfessorId,
    profilePictureBase64: request.profilePictureBase64 || undefined,
    active: true,
  });
  await ta.save();
  return ta.toJSON();
}

export async function taLogin(request) {
  const person = await User.findOne({ taId: request.username });
  if (!person || !(await bcrypt.compare(request.password, person.password))) {
    throw new Error('Invalid TA ID or password');
  }
  if (person.role !== 'TA') throw new Error('Not authorized as TA');

  const token = generateToken(person);
  return toAuthResponse(person, token);
}

export async function validateToken(token) {
  if (!token || !token.startsWith('TOKEN_')) throw new Error('Invalid token');
  const parts = token.split('_');
  if (parts.length < 3) throw new Error('Invalid token format');
  const userId = parts[1];
  const user = await User.findById(userId);
  if (!user) throw new Error('User not found');
  return user;
}
